File: backend/src/routes/watchlist-items.js
const express = require('express');
const requireAuth = require('../middleware/require-auth');
const state = require('../db/mock-db');

const router = express.Router();

function clone(value) {
  return JSON.parse(JSON.stringify(value));
}

function getNextId(table) {
  return state.nextId(table);
}

router.post('/watchlists/:watchlistId/items', requireAuth, (req, res) => {
  try {
    const userId = req.session.userId;
    const watchlistId = Number(req.params.watchlistId);

    const watchlist = state.watchlists.find(
      (w) => w.id === watchlistId && w.user_id === userId
    );

    if (!watchlist) {
      return res.status(404).json({
        success: false,
        error: 'Watchlist not found',
      });
    }

    const {
      option_contract_id,
      source_scan_id = null,
      status = 'active',
      entry_premium = null,
      entry_iv = null,
      entry_delta = null,
      entry_gamma = null,
      entry_vega = null,
      entry_theta = null,
      quantity = 1,
      notes = null,
    } = req.body;

    if (!option_contract_id) {
      return res.status(400).json({
        success: false,
        error: 'option_contract_id is required',
      });
    }

    const optionContract = state.option_contracts.find(
      (item) => item.id === Number(option_contract_id)
    );

    if (!optionContract) {
      return res.status(404).json({
        success: false,
        error: 'Option contract not found',
      });
    }

    const existing = state.watchlist_items.find(
      (item) =>
        item.watchlist_id === watchlistId &&
        item.option_contract_id === Number(option_contract_id)
    );

    if (existing) {
      return res.status(409).json({
        success: false,
        error: 'Option contract already in watchlist',
      });
    }

    const newItem = {
      id: getNextId('watchlist_items'),
      watchlist_id: watchlistId,
      option_contract_id: Number(option_contract_id),
      source_scan_id,
      status,
      entry_premium,
      entry_iv,
      entry_delta,
      entry_gamma,
      entry_vega,
      entry_theta,
      quantity: Number(quantity),
      notes,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    };

    state.watchlist_items.push(newItem);

    return res.status(201).json({
      success: true,
      watchlist_item: clone(newItem),
    });
  } catch (error) {
    console.log('WATCHLIST ITEMS POST ERROR:', error);

    return res.status(error.statusCode || 500).json({
      success: false,
      error: error.message || 'Internal server error',
    });
  }
});

router.get('/watchlists/:watchlistId/items', requireAuth, (req, res) => {
  try {
    const userId = req.session.userId;
    const watchlistId = Number(req.params.watchlistId);

    const watchlist = state.watchlists.find(
      (w) => w.id === watchlistId && w.user_id === userId
    );

    if (!watchlist) {
      return res.status(404).json({
        success: false,
        error: 'Watchlist not found',
      });
    }

    const items = state.watchlist_items.filter(
      (item) => item.watchlist_id === watchlistId
    );

    return res.status(200).json({
      success: true,
      watchlist_items: clone(items),
    });
  } catch (error) {
    console.log('WATCHLIST ITEMS GET ERROR:', error);

    return res.status(error.statusCode || 500).json({
      success: false,
      error: error.message || 'Internal server error',
    });
  }
});

router.delete('/watchlist-items/:itemId', requireAuth, (req, res) => {
  try {
    const userId = req.session.userId;
    const itemId = Number(req.params.itemId);

    const index = state.watchlist_items.findIndex((item) => {
      const watchlist = state.watchlists.find(
        (w) => w.id === item.watchlist_id && w.user_id === userId
      );
      return item.id === itemId && watchlist;
    });

    if (index === -1) {
      return res.status(404).json({
        success: false,
        error: 'Watchlist item not found',
      });
    }

    const [deleted] = state.watchlist_items.splice(index, 1);

    return res.status(200).json({
      success: true,
      deleted: clone(deleted),
    });
  } catch (error) {
    console.log('WATCHLIST ITEMS DELETE ERROR:', error);

    return res.status(error.statusCode || 500).json({
      success: false,
      error: error.message || 'Internal server error',
    });
  }
});

router.patch('/watchlist-items/:itemId/move', requireAuth, (req, res) => {
  try {
    const userId = req.session.userId;
    const itemId = Number(req.params.itemId);
    const { watchlist_id } = req.body;

    if (!watchlist_id) {
      return res.status(400).json({
        success: false,
        error: 'watchlist_id is required',
      });
    }

    const item = state.watchlist_items.find((entry) => entry.id === itemId);

    if (!item) {
      return res.status(404).json({
        success: false,
        error: 'Watchlist item not found',
      });
    }

    const targetWatchlist = state.watchlists.find(
      (w) => w.id === Number(watchlist_id) && w.user_id === userId
    );

    if (!targetWatchlist) {
      return res.status(404).json({
        success: false,
        error: 'Target watchlist not found',
      });
    }

    const sourceWatchlist = state.watchlists.find(
      (w) => w.id === item.watchlist_id && w.user_id === userId
    );

    if (!sourceWatchlist) {
      return res.status(403).json({
        success: false,
        error: 'Not allowed',
      });
    }

    item.watchlist_id = Number(watchlist_id);
    item.updated_at = new Date().toISOString();

    return res.status(200).json({
      success: true,
      watchlist_item: clone(item),
    });
  } catch (error) {
    console.log('WATCHLIST ITEMS MOVE ERROR:', error);

    return res.status(error.statusCode || 500).json({
      success: false,
      error: error.message || 'Internal server error',
    });
  }
});

module.exports = router;

File: backend-python/app/routers/watchlist_items.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session, joinedload

from app import models, schemas
from app.dependencies import get_current_user, get_db

router = APIRouter()


@router.get("/{watchlist_id}/items", response_model=List[schemas.WatchlistItemOut])
def list_items(
    watchlist_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    return (
        db.query(models.WatchlistItem)
        .options(joinedload(models.WatchlistItem.option_contract))
        .filter(models.WatchlistItem.watchlist_id == watchlist_id)
        .order_by(models.WatchlistItem.created_at.desc())
        .all()
    )


@router.get("/{watchlist_id}/items/{item_id}", response_model=schemas.WatchlistItemOut)
def get_item(
    watchlist_id: int,
    item_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    item = (
        db.query(models.WatchlistItem)
        .options(joinedload(models.WatchlistItem.option_contract))
        .filter(
            models.WatchlistItem.id == item_id,
            models.WatchlistItem.watchlist_id == watchlist_id,
        )
        .first()
    )
    if not item:
        raise HTTPException(status_code=404, detail="Watchlist item not found")
    return item


@router.post(
    "/{watchlist_id}/items/bulk-add",
    response_model=List[schemas.WatchlistItemOut],
    status_code=status.HTTP_201_CREATED,
)
def bulk_add_items(
    watchlist_id: int,
    body: schemas.BulkAddItems,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    created_ids = []

    for item_data in body.items:
        contract = _get_or_create_contract(db, item_data.option_contract)

        already_exists = (
            db.query(models.WatchlistItem)
            .filter(
                models.WatchlistItem.watchlist_id == watchlist_id,
                models.WatchlistItem.option_contract_id == contract.id,
            )
            .first()
        )
        if already_exists:
            continue

        item = models.WatchlistItem(
            watchlist_id=watchlist_id,
            option_contract_id=contract.id,
            source_scan_id=item_data.source_scan_id,
            entry_premium=item_data.entry_premium,
            entry_iv=item_data.entry_iv,
            entry_delta=item_data.entry_delta,
            entry_gamma=item_data.entry_gamma,
            entry_vega=item_data.entry_vega,
            entry_theta=item_data.entry_theta,
            quantity=item_data.quantity,
            notes=item_data.notes,
        )
        db.add(item)
        db.flush()
        created_ids.append(item.id)

    db.commit()

    if not created_ids:
        return []

    return (
        db.query(models.WatchlistItem)
        .options(joinedload(models.WatchlistItem.option_contract))
        .filter(models.WatchlistItem.id.in_(created_ids))
        .all()
    )


@router.patch("/{watchlist_id}/items/{item_id}", response_model=schemas.WatchlistItemOut)
def update_item(
    watchlist_id: int,
    item_id: int,
    body: schemas.WatchlistItemUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    item = _get_item_or_404(db, item_id, watchlist_id)

    for field, value in body.model_dump(exclude_none=True).items():
        setattr(item, field, value)

    db.commit()
    db.refresh(item)
    _ = item.option_contract
    return item


@router.delete("/{watchlist_id}/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
def remove_item(
    watchlist_id: int,
    item_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    item = _get_item_or_404(db, item_id, watchlist_id)
    db.delete(item)
    db.commit()


@router.post("/{watchlist_id}/items/bulk-delete", status_code=status.HTTP_204_NO_CONTENT)
def bulk_remove_items(
    watchlist_id: int,
    body: schemas.BulkRemoveItems,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    db.query(models.WatchlistItem).filter(
        models.WatchlistItem.id.in_(body.item_ids),
        models.WatchlistItem.watchlist_id == watchlist_id,
    ).delete(synchronize_session=False)
    db.commit()


@router.post("/{watchlist_id}/items/move", response_model=List[schemas.WatchlistItemOut])
def move_items(
    watchlist_id: int,
    body: schemas.MoveItems,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    _get_watchlist_or_404(db, body.target_watchlist_id, current_user.id)

    items = (
        db.query(models.WatchlistItem)
        .filter(
            models.WatchlistItem.id.in_(body.item_ids),
            models.WatchlistItem.watchlist_id == watchlist_id,
        )
        .all()
    )

    moved_ids = []
    for item in items:
        conflict = (
            db.query(models.WatchlistItem)
            .filter(
                models.WatchlistItem.watchlist_id == body.target_watchlist_id,
                models.WatchlistItem.option_contract_id == item.option_contract_id,
            )
            .first()
        )
        if not conflict:
            item.watchlist_id = body.target_watchlist_id
            moved_ids.append(item.id)

    db.commit()

    if not moved_ids:
        return []

    return (
        db.query(models.WatchlistItem)
        .options(joinedload(models.WatchlistItem.option_contract))
        .filter(models.WatchlistItem.id.in_(moved_ids))
        .all()
    )


# ── Alerts ────────────────────────────────────────────────────────────────────

@router.get("/{watchlist_id}/items/{item_id}/alerts", response_model=List[schemas.AlertOut])
def list_alerts(
    watchlist_id: int,
    item_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    _get_item_or_404(db, item_id, watchlist_id)
    return db.query(models.Alert).filter(models.Alert.watchlist_item_id == item_id).all()


@router.post(
    "/{watchlist_id}/items/{item_id}/alerts",
    response_model=schemas.AlertOut,
    status_code=status.HTTP_201_CREATED,
)
def create_alert(
    watchlist_id: int,
    item_id: int,
    body: schemas.AlertCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    _get_item_or_404(db, item_id, watchlist_id)
    alert = models.Alert(watchlist_item_id=item_id, **body.model_dump())
    db.add(alert)
    db.commit()
    db.refresh(alert)
    return alert


@router.patch("/{watchlist_id}/items/{item_id}/alerts/{alert_id}", response_model=schemas.AlertOut)
def update_alert(
    watchlist_id: int,
    item_id: int,
    alert_id: int,
    body: schemas.AlertUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    _get_item_or_404(db, item_id, watchlist_id)
    alert = (
        db.query(models.Alert)
        .filter(models.Alert.id == alert_id, models.Alert.watchlist_item_id == item_id)
        .first()
    )
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(alert, field, value)
    db.commit()
    db.refresh(alert)
    return alert


@router.delete(
    "/{watchlist_id}/items/{item_id}/alerts/{alert_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_alert(
    watchlist_id: int,
    item_id: int,
    alert_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    _get_watchlist_or_404(db, watchlist_id, current_user.id)
    _get_item_or_404(db, item_id, watchlist_id)
    alert = (
        db.query(models.Alert)
        .filter(models.Alert.id == alert_id, models.Alert.watchlist_item_id == item_id)
        .first()
    )
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    db.delete(alert)
    db.commit()


# ── Helpers ───────────────────────────────────────────────────────────────────

def _get_watchlist_or_404(db: Session, watchlist_id: int, user_id: int) -> models.Watchlist:
    wl = (
        db.query(models.Watchlist)
        .filter(models.Watchlist.id == watchlist_id, models.Watchlist.user_id == user_id)
        .first()
    )
    if not wl:
        raise HTTPException(status_code=404, detail="Watchlist not found")
    return wl


def _get_item_or_404(db: Session, item_id: int, watchlist_id: int) -> models.WatchlistItem:
    item = (
        db.query(models.WatchlistItem)
        .filter(
            models.WatchlistItem.id == item_id,
            models.WatchlistItem.watchlist_id == watchlist_id,
        )
        .first()
    )
    if not item:
        raise HTTPException(status_code=404, detail="Watchlist item not found")
    return item


def _get_or_create_contract(db: Session, data: schemas.OptionContractCreate) -> models.OptionContract:
    contract = (
        db.query(models.OptionContract)
        .filter(models.OptionContract.symbol_key == data.symbol_key)
        .first()
    )
    if not contract:
        contract = models.OptionContract(**data.model_dump())
        db.add(contract)
        db.flush()
    return contract
